heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java

```diff
@@ -198,65 +245,264 @@ public void close() {
+    for (int i = containers.size(); i <= numContainers - 1; i++) {
+      PackingUtils.allocateNewContainer(containers, containers.get(0).getCapacity(),
+          this.paddingPercentage);
+    }
+    int maxInstanceIndex = 0;
+    for (PackingPlan.ContainerPlan containerPlan : currentPackingPlan.getContainers()) {
+      for (PackingPlan.InstancePlan instancePlan : containerPlan.getInstances()) {
+        maxInstanceIndex = Math.max(maxInstanceIndex, instancePlan.getTaskId());
+      }
+    }
+    if (!componentsToScaleDown.isEmpty()) {
+      removeInstancesFromContainers(containers, allocation, componentsToScaleDown);
+    }
+    if (!componentsToScaleUp.isEmpty()) {
+      if (!assignInstancesToContainers(containers, allocation, componentsToScaleUp,
+          maxInstanceIndex + 1, "flexible")) {
+        //Not enough containers. Adjust the number of containers.
+        LOG.info(String.format("Increasing the number of containers to "
+            + "%s and attempting packing again.", this.numContainers + 1));
+        adjustNumContainers(1);
+        return Optional.absent();
+      }
+    }
+    PackingUtils.removeEmptyContainers(allocation);
+    return Optional.of(allocation);
+  }
+
+  /**
+   * Assigns instances to containers.
+   *
+   * @param containers helper data structure that describes the containers' status
+   * @param allocation existing packing plan
+   * @param parallelismMap component parallelism
+   * @param firstTaskIndex first taskId to use for the new instances
+   * @return true if the instances fit in the existing set of containers, false otherwise.
+   */
+  private boolean assignInstancesToContainers(
+      ArrayList<Container> containers, Map<Integer, List<InstanceId>> allocation,
+      Map<String, Integer> parallelismMap, int firstTaskIndex, String policyType) {
+    ArrayList<RamRequirement> ramRequirements = getRAMInstances(parallelismMap);
+    int globalTaskIndex = firstTaskIndex;
     int componentIndex = 0;
     for (String component : parallelismMap.keySet()) {
       long ramRequirement = ramRequirements.get(componentIndex).getRamRequirement();
       int numInstance = parallelismMap.get(component);
       for (int i = 0; i < numInstance; ++i) {
         Resource instanceResource = this.defaultInstanceResources.cloneWithRam(ramRequirement);
-        if (placeResourceCompliantRRInstance(containers, containerId,
-            new PackingPlan.InstancePlan(new InstanceId(component, globalTaskIndex, i),
-                instanceResource))) {
-          allocation.get(containerId).add(new InstanceId(component, globalTaskIndex, i));
-        } else {
-          //Automatically adjust the number of containers
-          increaseNumContainers();
-          LOG.info(String.format("Increasing the number of containers to "
-              + "%s and attempting packing again.", this.numContainers));
-          return null;
+        boolean sufficientNumContainers = true;
+        if ("strict".equals(policyType)) {
```

Comment:
better to represent PolicyType as an enum with a method that encapsulated this logic.